crates/commitlog/src/stream/writer.rs

```diff
@@ -107,36 +107,49 @@ where
         };
 
         let mut segment = repo.open_segment_writer(last)?;
-        let mut offset_index = repo::create_offset_index_writer(&repo, last, commitlog_options);
-        let meta = segment::Metadata::extract(last, &mut segment).or_else(|e| match e {
-            error::SegmentMetadata::InvalidCommit { sofar, source } => match on_trailing {
-                OnTrailingData::Error => Err(io::Error::new(io::ErrorKind::InvalidData, source)),
+        let offset_index = repo
+            .get_offset_index(last)
+            .inspect_err(|e| {
+                warn!("unable to open offset index for segment {last}: {e}");
+            })
+            .ok();
+
+        let (meta, offset_index) = match segment::Metadata::extract(last, &mut segment, offset_index.as_ref()) {
```

Comment:
I think this would be a bit easier to read if the conversion to `TxOffsetIndexMut` happened after the match, and the truncation would just borrow a mutable index like

```rust
if let Some(idx) = offset_index.as_mut().map(IndexFile::as_mut) {
// ...
}
```